Allow setters and getters to take additional arguments

nlupugla opened the pull request:
Please see the original pull request https://github.com/godotengine/godot/pull/76665 for summary of changes and discussion.

Edit 2023/Sept/18 8:43: Copied some of the examples over from the linked PR for convenience and expanded on them a bit.

### Examples

```gdscript
extends Node2D

signal sprite_property_set(property : StringName, value : Variant)
signal label_property_set(property : StringName, value : Variant)

@onready var sprite := $Icon as Sprite2D
@export var offset : Vector2 : set=set_in_sprite, get=get_in_sprite
@export var centered : bool : set=set_in_sprite, get=get_in_sprite

@onready var label := $Label as Label
@export var text : String : set=set_in_label, get=get_in_label
@export var language : String : set=set_in_label, get=get_in_label


func get_in_sprite(property : StringName) -> Variant:
    return sprite.get(property)
	
	
func set_in_sprite(property : StringName, value : Variant, emit_set := true) -> void:
    sprite.set(property, value)
    if emit_set: sprite_property_set.emit(property, value)


func get_in_label(property : StringName) -> Variant:
    return label.get(property)
	
	
func set_in_label(property : StringName, value : Variant, emit_set := true) -> void:
    label.set(property, value)
    if emit_set: label_property_set.emit(property, value)


func _ready() -> void:
    centered = true
    set(&"text", "The text was set without emitting a signal", false)
```

* Closes https://github.com/godotengine/godot-proposals/issues/6750.
* Closes https://github.com/godotengine/godot-proposals/issues/3096.
* Closes https://github.com/godotengine/godot-proposals/discussions/6107 (discussion).
* Closes https://github.com/godotengine/godot/issues/74144.

addresses the following proposals:
* https://github.com/godotengine/godot-proposals/issues/2817.
* https://github.com/godotengine/godot-proposals/issues/2983.
* https://github.com/godotengine/godot-proposals/issues/4867.
* https://github.com/godotengine/godot-proposals/issues/4868.

_Bugsquad edit:_ Add lists for issues/PRs links.

nlupugla committed:
Improve variable setters and getters
* Add support for name argument.
* Allow optional arguments.

Co-authored-by: dalexeev <danil@alexeev.xyz>

dalexeev commented:
1. You accidentally created a commit in which the author is akien-mga, and you are only the committer. Please fix this or the PR will not be counted as your contribution.
2. Copy "Closes ..." from the old PR to the first post.
3. I think the clarification is not needed in the title, you can add a link to the old PR in the first post.

nlupugla commented:
> 1. You accidentally created a commit in which the author is akien-mga, and you are only the committer. Please fix this or the PR will not be counted as your contribution.
> 2. Copy "Closes ..." from the old PR to the first post.
> 3. I think the clarification is not needed in the title, you can add a link to the old PR in the first post.

Done! Thanks for the help :)

dalexeev commented on modules/gdscript/gdscript.cpp:
```diff
@@ -1652,13 +1674,22 @@ bool GDScriptInstance::get(const StringName &p_name, Variant &r_ret) const {
-				r_ret = sptr->static_variables[E->value.index];
-				return true;
+				r_ret = sptr->static_variables[static_member->index];
+				return true; //index found
```
```suggestion
				return true;
```

A nitpick, but this comment seems useless to me. Or it should follow the guidelines (begin with a capital letter and end with a dot). Also in this PR there are other whitespace changes that are undesirable in my opinion. But again, this is a minor remark.

AThousandShips replied:
I'd agree that the removal of some empty lines are undesirable, probably considered useful for readability as they were kept when added originally

dalexeev commented on modules/gdscript/gdscript.cpp:
```diff
@@ -892,13 +891,19 @@ bool GDScript::_get(const StringName &p_name, Variant &r_ret) const {
 				if (E->value.getter) {
 					Callable::CallError ce;
-					r_ret = const_cast<GDScript *>(this)->callp(E->value.getter, nullptr, 0, ce);
+					const GDScript::MemberInfo *member = &E->value;
```
Below you use the name `static_member` for static variables. I suggest to do the same here, for consistency.

dalexeev commented on modules/gdscript/gdscript.cpp:
```diff
@@ -1627,15 +1642,22 @@ bool GDScriptInstance::get(const StringName &p_name, Variant &r_ret) const {
+				} else {
+					r_ret = const_cast<GDScriptInstance *>(this)->callp(member->getter, nullptr, 0, ce);
+				}
+				if (ce.error == Callable::CallError::CALL_OK) {
 					return true;
 				}
```
```suggestion
				return ce.error == Callable::CallError::CALL_OK;
```

I think it's more correct and above similar code. There is a code with the same error(?) below.

dalexeev replied:
> Hmm, your change makes sense to me, but I haven't thought too much about what returning false here would mean.

I haven't tested this, but I think there will be an error in the console if it returns `false`. In the current version, if the getter call fails, execution will continue and the raw value will be returned. I'm not sure if this is intentional.

> I've seen this pattern elsewhere in nearby code

> is it okay to do it in this PR or should it be separated?

Yes, other places should also be corrected. I think it can be added to this PR as it's a small change and not a big issue.

nlupugla replied:
I'm not so sure what the new behavior would be. There's a few cases where the return value of `get` is used, for instance: https://github.com/godotengine/godot/blob/cdd2313ba27d0a2600a18e849b4c5d1fd6a6e351/core/object/object.cpp#L456-L478. My thinking would be that this is enough of a behavior change that it makes sense to separate it, if only to make testing easier. However, you have much more experience on this than I do, so I'm happy to follow your lead :)

dalexeev replied:
If you have concerns about it, you don't have to fix it. I just drew attention to a possible error in the affected method.

nlupugla replied:
Sounds good. I'll leave it for another PR then.

In making these changes, I figured it would be cleaner to write a helper function like `call_getter/setter` to reduce code duplication. Sound reasonable?

dalexeev reviewed (commented):
1\. Here you also need to add the check.

https://github.com/godotengine/godot/blob/da4c810fcea9c2000e30e991d55b41328329efc6/modules/gdscript/gdscript.cpp#L1595-L1619

2\. After changing the error messages, the tests fail, you must also update the text in them.

<details>
<summary>Log</summary>

```
modules/gdscript/tests/gdscript_test_runner.cpp:213: ERROR: CHECK( result.passed ) is NOT correct!
  values: CHECK( false )
  logged: modules/gdscript/tests/scripts/analyzer/errors/setter_with_wrong_value_arg.gd
          GDTEST_ANALYZER_ERROR
Value argument of type "Array" in setter "set_named" is not compatible with a property of type "int".


modules/gdscript/tests/gdscript_test_runner.cpp:213: ERROR: CHECK( result.passed ) is NOT correct!
  values: CHECK( false )
  logged: modules/gdscript/tests/scripts/analyzer/errors/property_function_get_type_error.gd
          GDTEST_ANALYZER_ERROR
Return type of getter "get_prop" is not compatible with a property of type "int".


modules/gdscript/tests/gdscript_test_runner.cpp:213: ERROR: CHECK( result.passed ) is NOT correct!
  values: CHECK( false )
  logged: modules/gdscript/tests/scripts/analyzer/errors/property_function_set_type_error.gd
          GDTEST_ANALYZER_ERROR
Value argument of type "int" in setter "set_prop" is not compatible with a property of type "String".


modules/gdscript/tests/gdscript_test_runner.cpp:213: ERROR: CHECK( result.passed ) is NOT correct!
  values: CHECK( false )
  logged: modules/gdscript/tests/scripts/analyzer/errors/setter_with_wrong_name_arg.gd
          GDTEST_ANALYZER_ERROR
First mandatory argument of setter "set_named" with 2 mandatory arguments should be a property name and must be a String or StringName.
```

</details>

3\. Probably, tests should be duplicated for static variables, since their implementation is also largely duplicated. This does not please me, but I think it makes sense, taking into account point 1.

nlupugla commented:
I have some questions about a few details.

1. I just realized that in this version of the code, we don't seem to be allowing for extra optional arguments with getters/setters. For example, in the commit, line 903 of GDScript.cpp, we call `callp(member->getter, &args, 1, ce)` rather than, say `callp(member->getter, &args, args.size(), ce)`. Is this intentional? I suppose it is because `GDScript::set` only has 2 arguments. Should these be modified to have a variable number of optional args?
2. I'd like to introduce a helper function like `call_getter` and `call_setter` which would look roughly like
```
Callable::CallError GDScript::call_getter(const MemberInfo &p_member, const StringName &p_name, Variant &r_ret) const {
	Callable::CallError ce;
	if (p_member.getter_with_name) {
		Variant property_name = p_name;
		const Variant *args = &property_name;
		r_ret = const_cast<GDScript *>(this)->callp(p_member.getter, &args, 1, ce);
	} else {
		r_ret = const_cast<GDScript *>(this)->callp(p_member.getter, nullptr, 0, ce);
	}
	return ce;
}
```
Where should I stick this method in the class? Should it be private? public? protected? Do I need to make a copy of this method for `GDScriptInstance`?

dalexeev commented:
> 1. I just realized that in this version of the code, we don't seem to be allowing for extra optional arguments with getters/setters.

No, I think setters/getters with optional arguments are handled correctly. 1. Optional arguments are not required to be passed when called, so this works. 2. If you want to pass optional arguments, then you should call the setter/getter function instead:

```gdscript
my_var = value # The syntax itself doesn't allow additional arguments.
set("my_var", value) # The core doesn't support additional arguments.
set_my_var(value, opt_arg_1, opt_arg_2) # It works.
```

In fact, #74144 reports the issue that you cannot use such functions as setters/getters. But we're not talking about setters/getters supporting extra arguments. Just now you can combine these two things in one function.

> 2\. I'd like to introduce a helper function like `call_getter` and `call_setter`

Yes, for GDScriptInstance this should be a separate method, because for non-static variables there is a difference on which instance the setter/getter method is called (for static variables, the method is called on the script, not on the instance). Alternatively, you can use a macro. I'm not sure if we need to follow DRY in this case, but let's try it and see what happens.

nlupugla commented:
That makes total sense about the optional arguments, thanks for explaining that :)

I'll be honest, I'm kind of scared of macros, so I'll just stick to functions :). One thing I thought of for the `GDScriptInstance` helpers was to call `script.ptr()->call_getter`, but when I compared the code for `GDScript::callp` and `GDScriptInstance::callp`, they looked to have some non-trivial differences. For the modifiers (eg: public/protected/private, const, virtual, etc), I'm planning to just copy whatever I see in `callp`. I'll write the methods directly above wherever `callp` appears in the source. Finally, for an internal-use-only function like `call_getter`, is it more appropriate to name it `_call_getter` to signify that it's private, or would that imply that it is intended to be overwritten?

Thanks!

nlupugla commented:
Apologies for the multiple commits. I'm working on rebasing them into 1 commit right now.

nlupugla commented:
Some of my changes got screwed up when I squashed. Thankfully I made a backup branch :) This last commit should be good to go now!

nlupugla commented:
My bad, forgot to add tests for static variables. I'll write those now.

nlupugla commented:
I ran the GDScript tests properly now instead of manually filling in the .out files, so the CI should no longer complain about getting the wrong error messages in the test with this latest commit.

AThousandShips commented on modules/gdscript/tests/scripts/runtime/features/setter_getter_with_name.gd:
```diff
@@ -0,0 +1,99 @@
+
+static func static_get_named_typed(name: StringName, optional: bool = true) -> int:
+	prints("static_get_named_typed", name, optional)
+	return 0
```
```suggestion
	return 0

```
Missing termination, not enforced but neater with it

dalexeev replied:
Also this test is hard to read in my opinion.
@nlupugla I prepared another test, let me know what you think about it: [gist](https://gist.github.com/dalexeev/a694f8306d1ecdf429ad8769eb105048).

nlupugla replied:
I like that layout much better! I was thinking that the test was hard to read, which is why I changed the variable names from "a, b, c, d" to something more descriptive. I'll commit the test later today.

dalexeev commented on modules/gdscript/gdscript.h:
```diff
@@ -325,6 +329,8 @@ class GDScriptInstance : public ScriptInstance {
 	virtual void get_method_list(List<MethodInfo> *p_list) const;
 	virtual bool has_method(const StringName &p_method) const;
+	Callable::CallError call_getter(const GDScript::MemberInfo &p_member_info, const StringName &p_name, Variant &r_ret) const;
+	Callable::CallError call_setter(const GDScript::MemberInfo &p_member_info, const StringName &p_name, const Variant &p_value);
```
Note that you have added these methods to `protected` section of `GDScript` , to `public` section of `GDScriptInstance`, and the `MemberInfo` struct is `private`. I see that you did this so that the methods are next to `callp()`, but I think they have a different purpose. In my opinion `call_getter()`/`call_setter()` methods should be private.

Or maybe we should discard them, since there are only 3 calls for 2 declarations and 2 definitions. This is not a lot of code duplication and yet new uses of these methods are not expected, I think.

nlupugla replied:
I agree about making the methods private or protected. I'm never really sure when something should be protected versus private, but I'll take your word that private is more suitable in this case.

I'd prefer to keep the helper methods as I think it makes the code more readable and easier to maintain. My general philosophy is that if a comment was needed to explain a block of code , the block should be rolled into a function with a self-explanatory name (big exceptions being for comments that explain some kind of non-obvious implementation or cite a source). That said you have much more experience maintaining big code, so I'm happy to defer to you if you feel strongly.

dalexeev replied:
I do not insist on code duplication here. It's just my personal opinion in the particular case, I'd like to know what others think about it.

Regarding the visibility modifiers, I explained the problem: the visibility levels are different. In my opinion it is wrong that a public/protected method has a parameter of a private type (`MemberInfo` struct).

dalexeev commented on modules/gdscript/gdscript.cpp:
```diff
@@ -892,14 +916,15 @@ bool GDScript::_get(const StringName &p_name, Variant &r_ret) const {
 				return true;
 			}
 		}
-
 		{
 			HashMap<StringName, MemberInfo>::ConstIterator E = top->static_variables_indices.find(p_name);
 			if (E) {
 				if (E->value.getter) {
-					Callable::CallError ce;
-					r_ret = const_cast<GDScript *>(this)->callp(E->value.getter, nullptr, 0, ce);
-					return true;
+					const GDScript::MemberInfo *member = &E->value;
+					Callable::CallError ce = call_getter(*member, p_name, r_ret);
+					if (ce.error == Callable::CallError::CALL_OK) {
+						return true; //TODO: should this just be return ce.error == Callable::CallError::CALL_OK?
+					}
```
Note that this _already_ makes a change to the logic. Also, the comment is missing a space before "TODO". The same code is below in 5 more places.

nlupugla replied:
This particular line was kind of odd. It was the only getter/setter call that didn't use the call error and just returned true no matter what. At first, I did the same thing, but the CI complained that I didn't use the returned call error. I would love if someone could explain the logic around this particular get call

dalexeev reviewed (commented):
* Some of the old comments (about the `static_member` name, about the removal of blank lines) have not yet been addressed.
* There also seems to be a lack of error tests for too many or too few mandatory parameters of setter/getter functions.

nlupugla commented:
> * Some of the old comments (about the `static_member` name, about the removal of blank lines) have not yet been addressed.
> * There also seems to be a lack of error tests for too many or too few mandatory parameters of setter/getter functions.

Thanks for the review again! Sorry about forgetting those old comments. I'll try getting into the habit of leaving Todos for myself as soon as I read a review.

Good point about the error tests, I'll add those in.

nlupugla commented:
I think I've addressed all the comments now except for the minor change in logic related to the getter always returning true versus only returning true if there was no call error.

dalexeev commented on modules/gdscript/gdscript_compiler.cpp:
```diff
@@ -1139,7 +1148,7 @@ GDScriptCodeGenerator::Address GDScriptCompiler::_parse_expression(CodeGen &code
 				if (!known_type || !is_shared) {
 					// If this is a class member property, also assign to it.
-					// This allow things like: position.x += 2.0
+					// This allows things like: position.x += 2.0
```
```suggestion
					// This allows things like: position.x += 2.0.
```

dalexeev commented on modules/gdscript/gdscript_compiler.cpp:
```diff
@@ -1154,7 +1163,11 @@ GDScriptCodeGenerator::Address GDScriptCompiler::_parse_expression(CodeGen &code
 							gen->write_jump_if_shared(assigned);
 						}
 						if (member_property_has_setter && !member_property_is_in_setter) {
+							// Call setter
```
```suggestion
							// Call setter.
```

nlupugla commented on modules/gdscript/gdscript_analyzer.cpp:
```diff
@@ -1291,15 +1291,25 @@ void GDScriptAnalyzer::resolve_class_body(GDScriptParser::ClassNode *p_class, co
 
-						if (getter_function->parameters.size() != 0 || return_datatype.has_no_type()) {
-							push_error(vformat(R"(Function "%s" cannot be used as getter because of its signature.)", getter_function->identifier->name), member.variable);
+						if (return_datatype.has_no_type()) {
+							push_error(vformat(R"(Getter "%s" must return a value.)", getter_function->identifier->name), member.variable);
```
Good point! I'll remove that logic.

nlupugla commented on modules/gdscript/gdscript_analyzer.cpp:
```diff
@@ -1291,15 +1291,25 @@ void GDScriptAnalyzer::resolve_class_body(GDScriptParser::ClassNode *p_class, co
 						} else if (!is_type_compatible(member.variable->datatype, return_datatype, true)) {
-							push_error(vformat(R"(Function with return type "%s" cannot be used as getter for a property of type "%s".)", return_datatype.to_string(), member.variable->datatype.to_string()), member.variable);
-
+							push_error(vformat(R"(Return type of getter "%s" is not compatible with a property of type "%s".)", getter_function->identifier->name, return_datatype.to_string()), member.variable);
```
That makes sense, although your suggested change confuses me. Isn't your suggestion longer than the original? Did you mean to remove the names of the getter and property? Perhaps something like below?
```cpp
vformat(R"(Return type %s of getter is not compatible with value type %s)", return_datatype.to_string(), member.variable->datatype.to_string())
```

nlupugla commented on modules/gdscript/gdscript_analyzer.cpp:
```diff
@@ -1291,15 +1291,25 @@ void GDScriptAnalyzer::resolve_class_body(GDScriptParser::ClassNode *p_class, co
+								GDScriptParser::DataType name_arg_type = getter_function->parameters[0]->get_datatype();
+								if (!name_arg_type.is_variant() && name_arg_type.is_hard_type()) {
+									if (name_arg_type.kind != GDScriptParser::DataType::BUILTIN || (name_arg_type.builtin_type != Variant::STRING && name_arg_type.builtin_type != Variant::STRING_NAME)) {
+										push_error(vformat(R"(First mandatory argument of getter "%s" should be a property name and must be a String or StringName.)", getter_function->identifier->name), member.variable);
```
Makes sense!

nlupugla commented on modules/gdscript/gdscript_analyzer.cpp:
```diff
@@ -1314,27 +1324,38 @@ void GDScriptAnalyzer::resolve_class_body(GDScriptParser::ClassNode *p_class, co
+						} else {
+							GDScriptParser::ParameterNode *value_arg = setter_function->parameters[setter_function->mandatory_parameter_count - 1];
+							if (!is_type_compatible(member.variable->datatype, value_arg->datatype, true)) {
+								push_error(vformat(R"(Value argument of type "%s" in setter "%s" is not compatible with a property of type "%s".)", value_arg->datatype.to_string(), setter_function->identifier->name, member.variable->datatype.to_string()), member.variable);
```
This message should probably be reworded to be shorter as you suggested above.

nlupugla commented:
Finally had some time to make the requested changes! Hopefully the PR is ready to ship now :)

nlupugla commented:
Darn, co close to passing the CI...

I don't really understand the error message. Any tips on how to fix?

AThousandShips commented:
It's a sporadic error, unrelated to PR

nlupugla commented:
Passed!! 🥳

nlupugla commented on modules/gdscript/gdscript.cpp:
```diff
@@ -938,32 +964,30 @@ bool GDScript::_set(const StringName &p_name, const Variant &p_value) {
+				Callable::CallError ce = call_setter(*static_member, p_name, value);
+				if (ce.error == Callable::CallError::CALL_OK) {
+					return true; // TODO: should this just be return ce.error == Callable::CallError::CALL_OK?
+				};
```
Fixed it! Surprised that wasn't caught by the formatter.

AThousandShips replied:
It's not technically wrong, but isn't correct either

dalexeev commented on modules/gdscript/gdscript_analyzer.cpp:
```diff
@@ -1335,15 +1335,23 @@ void GDScriptAnalyzer::resolve_class_body(GDScriptParser::ClassNode *p_class, co
-						if (getter_function->parameters.size() != 0 || return_datatype.has_no_type()) {
-							push_error(vformat(R"(Function "%s" cannot be used as getter because of its signature.)", getter_function->identifier->name), member.variable);
+						if (getter_function->mandatory_parameter_count > 1) {
+							push_error(vformat(R"(Getters cannot have more than 1 mandatory argument.)"), member.variable);
```
```suggestion
							push_error(R"(A getter cannot have more than 1 mandatory argument.)", member.variable);
```

dalexeev commented on modules/gdscript/gdscript_analyzer.cpp:
```diff
@@ -1335,15 +1335,23 @@ void GDScriptAnalyzer::resolve_class_body(GDScriptParser::ClassNode *p_class, co
+								GDScriptParser::DataType name_arg_type = getter_function->parameters[0]->get_datatype();
+								if (!name_arg_type.is_variant() && name_arg_type.is_hard_type()) {
+									if (name_arg_type.kind != GDScriptParser::DataType::BUILTIN || (name_arg_type.builtin_type != Variant::STRING && name_arg_type.builtin_type != Variant::STRING_NAME)) {
+										push_error(vformat(R"(First mandatory argument of getters must be a property name of type String or StringName.)"), member.variable);
```
```suggestion
										push_error(R"(First mandatory argument of a getter must be String or StringName.)", member.variable);
```

dalexeev commented on modules/gdscript/gdscript_analyzer.cpp:
```diff
@@ -1335,15 +1335,23 @@ void GDScriptAnalyzer::resolve_class_body(GDScriptParser::ClassNode *p_class, co
 						} else if (!is_type_compatible(member.variable->datatype, return_datatype, true)) {
-							push_error(vformat(R"(Function with return type "%s" cannot be used as getter for a property of type "%s".)", return_datatype.to_string(), member.variable->datatype.to_string()), member.variable);
-
+							push_error(vformat(R"(Return type "%s" of getter is not compatible with value type "%s".)", return_datatype.to_string(), member.variable->datatype.to_string()), member.variable);
```
```suggestion
							push_error(vformat(R"(Getter return type "%s" is not compatible with variable type "%s".)", return_datatype.to_string(), member.variable->datatype.to_string()), member.variable);
```

dalexeev commented on modules/gdscript/gdscript_analyzer.cpp:
```diff
@@ -1358,27 +1366,38 @@ void GDScriptAnalyzer::resolve_class_body(GDScriptParser::ClassNode *p_class, co
+						} else {
+							GDScriptParser::ParameterNode *value_arg = setter_function->parameters[setter_function->mandatory_parameter_count - 1];
+							if (!is_type_compatible(member.variable->datatype, value_arg->datatype, true)) {
+								push_error(vformat(R"(Value argument of type "%s" is not compatible with value type "%s".)", value_arg->datatype.to_string(), member.variable->datatype.to_string()), member.variable);
```
```suggestion
								push_error(vformat(R"(Value argument of type "%s" is not compatible with variable type "%s".)", value_arg->datatype.to_string(), member.variable->datatype.to_string()), member.variable);
```

Also I really don't like the "value argument" even though I originally suggested it. Maybe we should just specify the argument name despite the long error message?

nlupugla replied:
That makes sense, considering it probably won't be clear what we mean by "value argument" and in a setter, it could be the first or second argument depending on whether it's a "named" setter or not.

I'll go with "Setter argument of type "%s" is not compatible with variable type "%s".

dalexeev commented on modules/gdscript/gdscript_analyzer.cpp:
```diff
@@ -1358,27 +1366,38 @@ void GDScriptAnalyzer::resolve_class_body(GDScriptParser::ClassNode *p_class, co
+									GDScriptParser::DataType name_arg_type = setter_function->parameters[0]->datatype;
+									if (!name_arg_type.is_variant() && name_arg_type.is_hard_type()) {
+										if (name_arg_type.kind != GDScriptParser::DataType::BUILTIN || (name_arg_type.builtin_type != Variant::STRING && name_arg_type.builtin_type != Variant::STRING_NAME)) {
+											push_error(vformat(R"(First mandatory argument of setters with 2 mandatory arguments must be a property name of type String or StringName.)"), member.variable);
```
```suggestion
											push_error(R"(First mandatory argument of a setter with 2 mandatory arguments must be String or StringName.)", member.variable);
```

nlupugla replied:
Agree mostly here, but I prefer "must be of type String or StringName" to "must be String or StringName" because the latter could be interpreted as meaning the argument literally has to be one of two values: `String` or `StringName`.

nlupugla commented on modules/gdscript/gdscript.cpp:
```diff
@@ -896,14 +921,15 @@ bool GDScript::_get(const StringName &p_name, Variant &r_ret) const {
-					return true;
+					const GDScript::MemberInfo *static_member = &E->value;
+					Callable::CallError ce = call_getter(*static_member, p_name, r_ret);
+					if (ce.error == Callable::CallError::CALL_OK) {
+						return true; // TODO: should this just be return ce.error == Callable::CallError::CALL_OK?
+					}
```
Makes sense to me. The only problem is that the CI will complain that the call error `ce` is not used. The CI didn't catch this before because `ce` was filled via `call_p`, but now it is returned by my `call_getter` helper. I could just not use my helper function here, but I'm not sure that's the best approach.

dalexeev replied:
You could try just calling `call_getter()` without assigning its return value to an unused variable. As far as I remember, the warning about discarding the returned value is not enabled.

dalexeev reviewed (commented):
The code looks good to me (but I haven't tested it again). There are only 2 points left in which we cannot reach consensus over several iterations:

1. Error messages. Don't underestimate this, they are important to users. I think my suggestions are consistent, but you don't apply them consistently.
2. Inconsistent error handling when calling setters/getters in `gdscript.cpp`. I'm not asking you to fix it in this PR, you could leave it as it is if you didn't change the logic (by wrapping `return` into `if`). But since you're changing this, I suggest an option that's closer to #81139.

I really like this PR. I hope we reach a consensus soon!

nlupugla commented on modules/gdscript/gdscript_analyzer.cpp:
```diff
@@ -1358,27 +1366,38 @@ void GDScriptAnalyzer::resolve_class_body(GDScriptParser::ClassNode *p_class, co
+						} else {
+							GDScriptParser::ParameterNode *value_arg = setter_function->parameters[setter_function->mandatory_parameter_count - 1];
+							if (!is_type_compatible(member.variable->datatype, value_arg->datatype, true)) {
+								push_error(vformat(R"(Setter argument "%s" is not compatible with variable type "%s".)", value_arg->identifier->name, member.variable->datatype.to_string()), member.variable);
```
Lol, I think we've come full circle :)

I think that's how I wrote the error message before, but we wanted the error message to be shorter.

How about `Setter argument "%s" of type "%s" is not compatible with variable type  "%s"`?

Side note: should I change all the instances of "not compatible" to "incompatible" to save a few characters?

nlupugla commented:
That is a really good point dalexeev. Let's keep the signature as is then. Less work for me too :)

MarianoGnu commented:
I personally don't like how this feature is meant to work from UX perspective, i dont think the amount of boilerplate is reduced.
i think a syntax i would expect is:
```
@export_proxy_property(NodePath("ChildSprite:centered")) var sprite_centered_or_any_name = true
@export_proxy_property(NodePath"ChildSprite:self_modulate:a") var sprite_opacity = 0.5
```
and then you dont need any setter or getter at all

nlupugla commented:
> I personally don't like how this feature is meant to work from UX perspective, i dont think the amount of boilerplate is reduced. i think a syntax i would expect is:
> 
> ```
> @export_proxy_property(NodePath("ChildSprite:centered")) var sprite_centered_or_any_name = true
> @export_proxy_property(NodePath("ChildSprite:self_modulate:a")) var sprite_opacity = 0.5
> ```
> 
> and then you dont need any setter or getter at all

Thanks for the comment MarianoGnu! Your idea is very similar to what I originally proposed a few months ago: https://github.com/godotengine/godot-proposals/issues/6750. The current approach is a compromise that doesn't achieve quite the same level of boilerplate reduction, but is also much more flexible and aligned with GDScript's current syntax. For example, what if you want the proxy property to emit a signal when it changes but the original property doesn't emit a signal? The @export_proxy_property approach doesn't provide a clean way to do this, whereas the setter/getter approach does.
